pkg/sql/catalog/descpb/structured.proto

```diff
@@ -1473,3 +1480,16 @@ message Descriptor {
     SchemaDescriptor schema = 4;
   }
 }
+
+// Wrapper for the bytes of the value of an index that also contains bit for
+// whether or not the value was deleted. A wrapper of arbitrary index values
+// with the additional delete bit was chosen over a separate index encoding type
+// because there would have to be a separate type for each encoding that we
+// already have for indexes. The alternative would get harder to maintain if we
+// added more index encodings in the future.
+message IndexValueWrapper {
```

Comment:
Why is this in the `descpb` package? Rather, my real question is, why do we need a protobuf message to wrap an encoded index value at all? 

It seems that the same results could be achieved by, say, pre-pending a 0 or a 1 byte to the encoded value bytes, with 0 being a tombstone and 1 signalling an actual value. Conceptually, it's the same thing as what you're proposing here, unless I'm missing something. 

Perhaps I'm way off base here but in fact it seems to me that there might be a way to achieve what you want to do without needing to introduce a new encoding format at all, by going deeper into the workings of `rowenc` into the messy innards of `EncodePrimaryIndex` and `EncodeSecondaryIndex`. These all use the same `writeColumnValues` function to encode non-key column values: 
https://github.com/cockroachdb/cockroach/blob/48a8dd8eb4990fdf20959cf0cb50a3d723922c42/pkg/sql/rowenc/index_encoding.go#L1438-L1461

This admittedly is not my area of expertise but couldn't we pretend that every table has a secret "deletion tombstone" nullable integer column that has a very large, out-of-bounds ID? Marking a row as deleted would be equivalent to having a non-NULL value in that column. 

Encoding a deletion tombstone would involve an extra iteration of the for-loop in that function with `colIDDiff` set to `math.MaxUint`: column IDs start at 1 and end at `math.MaxUint` so `colIDDiff` can in practice never exceed  `math.MaxUint -1`, making this a legal value (from the point of view of the encoding) but impossible for a real column to have. 

Similarly, when decoding in `DecodeRow` we check `colIDDiff == math.MaxUint` in which case the row is considered deleted.

Is this crazy? If not, implementing this properly would involve a bit of plumbing, however there's no need to modify the index descriptor at all, existing rows all remain valid, etc.

Reply:
Hm.

>  pre-pending a 0 or a 1 byte to the encoded value bytes, with 0 being a tombstone and 1 signalling an actual value

Eh, I don't love rolling out own "pack fields into bytes" encoding when that's exactly what proto the proto will do for us -- prepend the field ID, and then its value, but using a mature, well tested library.

I'm a little wary of the magic extra column, and how closely it starts to tie this to the internals of SQL value encoding. I think I sorta like that we just implement the Put/Del API but take whatever was put, or del'ed, and record it according to our own container, with that fact separated from the details of what was put or del'ed.

Reply:
I'm not saying we should pre-pend with magic numbers, I'm saying that we're essentially doing the same thing with the protos. But then I didn't realize there was a desire to keep this wrapper at a different abstraction level. Again, the rationale seems like it's worth explaining in comments.

I trust your judgment about this and defer to you. That being said I'd prefer if this wrapper proto lived somewhere else; as I understand things, it doesn't belong in descpb.

Reply:
Yeah, fair, descpb is a little odd since this is row data not descriptor metadata. I think it moved here since @tbg asked for it -- as a sql-layer-specific type -- to not be in roachpb. We might be able sell @tbg on the idea that with a message type with more generic name, like "DeleteableValue", a message that has a generic bytes value and deleted bit does make sense in roachpb, or maybe we could just move the message with its current index-specific name over to execinfrapb?